projects/libhtp/fuzz.diff

```diff
@@ -0,0 +1,294 @@
+ test_all_LDADD = libgtest.a -lpthread $(LDADD)
+ 
++test_fuzz_SOURCES = fuzz/onefile.c fuzz/fuzz_htp.c test.c
++test_fuzz_LDADD = $(LDADD)
++
+ libgtest_a_SOURCES = gtest/gtest-all.cc gtest/gtest_main.cc gtest/gtest.h
+ 
+ TESTS_ENVIRONMENT = srcdir=$(srcdir)/files
+diff --git a/test/fuzz/fuzz_htp.c b/test/fuzz/fuzz_htp.c
+new file mode 100644
+index 0000000..b9c274a
+--- /dev/null
++++ b/test/fuzz/fuzz_htp.c
+@@ -0,0 +1,157 @@
++/**
++ * @file
++ * @author Philippe Antoine <contact@catenacyber.fr>
++ * fuzz harness for libhtp
++ */
++
++
++#include <errno.h>
++#include <stdlib.h>
++#include <sys/types.h>
++#include <string.h>
++#include <stdio.h>
++
++#include <sys/stat.h>
++#include <fcntl.h>
++
++#include "htp/htp.h"
++#include "test/test.h"
++
++
++FILE * logfile = NULL;
++
++
++/**
++ * Invoked at the end of every transaction. 
++ *
++ * @param[in] connp
++ */
++static int callback_response(htp_tx_t *out_tx) {
++    if (out_tx != NULL) {
++        char *x = bstr_util_strdup_to_c(out_tx->request_line);
++        fprintf(logfile, "%s\n", x);
++        free(x);
++    }
++    return 0;
++}
++
++/**
++ * Invoked every time LibHTP wants to log. 
++ *
++ * @param[in] log
++ */
++static int callback_log(htp_log_t *log) {
++    fprintf(logfile, "[%d][code %d][file %s][line %d] %s\n",
++        log->level, log->code, log->file, log->line, log->msg);
++    return 0;
++}
++
++void fuzz_openFile(const char * name) {
++    if (logfile != NULL) {
++        fclose(logfile);
++    }
++    logfile = fopen(name, "w");
++}
++
++int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
++    htp_cfg_t *cfg;
++    htp_connp_t * connp;
++    int rc;
++    test_t test;
++
++    //initialize output file
++    if (logfile == NULL) {
++        logfile = fopen("/dev/null", "w");
++        if (logfile == NULL) {
++            return 0;
++        }
++    }
++
++    // Create LibHTP configuration
++    cfg = htp_config_create();
++    if (htp_config_set_server_personality(cfg, HTP_SERVER_IDS) != HTP_OK) {
++        htp_config_destroy(cfg);
++        return 0;
++    }
++    htp_config_register_response_complete(cfg, callback_response);
++    htp_config_register_log(cfg, callback_log);
++
++    connp = htp_connp_create(cfg);
++    htp_connp_set_user_data(connp, (void *) 0x02);
++    htp_connp_open(connp, (const char *) "192.168.2.3", 12345, (const char *) "192.168.2.2", 80, NULL);
++
++    test.buf = Data;
++    test.len = Size;
++    test.pos = 0;
++    test.chunk = NULL;
++
++    // Find all chunks and feed them to the parser
++    int in_data_other = 0;
++    char *in_data = NULL;
++    size_t in_data_len = 0;
++    size_t in_data_offset = 0;
++    int out_data_other = 0;
++    char *out_data = NULL;
++    size_t out_data_len = 0;
++    size_t out_data_offset = 0;
++
++    for (;;) {
++        if (test_next_chunk(&test) <= 0) {
++            break;
++        }
++        if (test.chunk_direction == CLIENT) {
```

Comment:
Is `chunk_direction` properly initialized?

Reply:
It should be initialized by `test_next_chunk`
If it is not, it is a bug that we should fix in `test_next_chunk`